### ISSUE TITLE
support cert-manager to issue certs

### ISSUE THREAD

**Issue by raoofm** [banzaicloud/bank-vaults#354]
From slack
**Additional context**

Raoof   [Yesterday at 6:02 PM]
@nandi is there a plan to support cert-manager to issue certs
35 replies

Raoof   [17 hours ago]
actually i could do it by having a yaml in my helm chart say

Raoof   [17 hours ago]
apiVersion: certmanager.k8s.io/v1alpha1
kind: Certificate

Raoof   [17 hours ago]
but if vault-operator starts earlier then it wont wait for these secrets to be generated and mounted and will create the self signed

Raoof   [17 hours ago]
probably there is manual step required to do that

jurgen   [17 hours ago]
yeah, well. how do you plan to do this?

jurgen   [17 hours ago]
I created an ingress to handle external connections

jurgen   [17 hours ago]
but you can not use cert-manager for the TLS on the vault binary…. how can cert-manager validate that?

Raoof   [17 hours ago]
I mean I’ll create the cert as secret before hand and then deploy operator

jurgen   [17 hours ago]
right

Raoof   [17 hours ago]
which will check if the tls certs are present it will not create

Raoof   [17 hours ago]
and just start vault by mounting those certs

jurgen   [17 hours ago]
yeah, but for what hostname?

Raoof   [17 hours ago]
oh you mean the self signed ones have like vault.namespace and localhost

jurgen   [17 hours ago]
yeah, so if you cert-manager/LE… what hostname? That is the bit that gets me.. It makes sense to use the generated self signed internally in the cluster and then an ingress for external.. I think. (edited)

Raoof   [17 hours ago]
yeah i’m on the same page

Raoof   [17 hours ago]
not sure, right now cert-manager created certs are being used by me at nginx and self signed ones by vault

jurgen   [17 hours ago]
unless you can do `vault`, `vault.namespace` and `vault.example.com` in one cert but I don’t think you can. LE recommends for things like`vault` and `vault.namespace` you should use your own CA.

jurgen   [17 hours ago]
yeah, cool. As you said, same page.

Raoof   [17 hours ago]
but that is not end to end tls and fails compliance

jurgen   [17 hours ago]
I use ingress nginx and you can have TLS backends… so TLS from external to ingress and then TLS from the ingress to vault.

jurgen   [17 hours ago]
would that still cover teh compliance?

Raoof   [17 hours ago]
no

jurgen   [17 hours ago]
annoying, can not reverse proxy?

Raoof   [17 hours ago]
as the sessions get decrypted and then encrypted

jurgen   [17 hours ago]
right

Raoof   [16 hours ago]
that means eavesdropping could happen…

Raoof   [16 hours ago]
end to end tls mandates no decryption

jurgen   [16 hours ago]
yeah, I hear you

jurgen   [16 hours ago]
ok

Raoof   [16 hours ago]
i use kubernetes nginx ingress as well

Raoof   [16 hours ago]
and use tls backend too

Raoof   [16 hours ago]
ok

Raoof   [16 hours ago]
I’ll keep the discussion open for @nandi

jurgen   [16 hours ago]
yeah, good chat. Very interesting.

nandi   [5 minutes ago]
Yeah good idea, please post it as a github issue, as otherwise it can easily get lost here :slightly_smiling_face:

**Comment by bonifaido** [comment]
AutoCert is also a nice cert management system: https://github.com/smallstep/certificates/blob/master/autocert/README.md

**Comment by raoofm** [comment]
yeah autocert is a good option for within cluster but for vault you will have scenarios where applications outside your cluster would access too. Even the ops guys to use the UI or cli from outside kubernetes.

**Comment by primeroz** [comment]
Cert-manager is now supported as describe in https://github.com/banzaicloud/bank-vaults/blob/master/operator/deploy/cr-customports.yaml 

Right now though there is no automatic restart of vault when cert-manager update the secret. 
Once that is fixed by #380 this issue should probably done.

**Comment by primeroz** [comment]
Now that #380 is implemented i think we have all the building blocks to use cert-manager for external listeners.

Internal listener is still handled as before and work to implement renewal of the internal cert is underway

Should this be closed ?

**Comment by bonifaido** [comment]
I would be very happy if we could create a CR example (if possible) with the cert manager use-case, but if no we can close it anyhow. :)

**Comment by primeroz** [comment]
this one [here](https://github.com/banzaicloud/bank-vaults/blob/master/operator/deploy/cr-customports.yaml) has all the bits for the vault spec needed by the cert-manager, is only missing the cert-manager configuration , but basically all you need is :
1. Cert-manager and certificate request ( out of scope )
1. Volumes and VolumeMounts in the Vault Spec for the extra secret
1. Custom Ports defined for the extra listener
1. Extra Listener for the external instance using the mounted secret
1. watchedSecretsLabels to trigger restart of statefulset when letsencrypt cert is updated 
1. An Ingress with TLS Passthrough ( which we are not defining here in the spec, i personally manage that ingress by myself, i could look into it for the example )

**Comment by bonifaido** [comment]
Don't worry, I will put together an example, and will fix the Ingress creation by the operator.

**Comment by kunickiaj** [comment]
Great, so is there any work left on this? Just got bit where I had not seen this and re-used the vault-tls secret and placed my own certs in it.. Got a nasty surprise when the operator regenerated self signed certs when mine were nearing expiration and caused all the clients connecting to Vault to break.

Our use case is that, we want cross-cluster (or from within VPC) access to Vault running with the operator. We're using a private LoadBalancer type service for exposing it to outside clients.

We're using Vault itself to issue its own cert after bootstrapping (using a CA managed outside of Vault). Was able to do this with vault.vault and 127.0.0.1 as SAN and IP SAN in addition to the FQDN exposed over the private LoadBalancer.

The above solution with cert-manager seems like a much nicer option, since we hadn't yet solved the automatic renewal part.

**Comment by kunickiaj** [comment]
Tried this out, and it seems that you _must_ use a separate non-default port for this to work. Otherwise, I've ended up with Vault going down because the operator is still refreshing the vault-tls secret with its own self-signed one.

I've also tried using cert manager to set vault-tls instead of a different secret, while vault comes up, the operator no longer works properly. Guessing this is likely due to the fact that cert-manager uses tls.crt, tls.key whereas the operator is looking for server.crt server.key in the secret.

```
{"level":"error","ts":1568783301.406511,"logger":"kubebuilder.controller","msg":"Reconciler error","controller":"vault-controller","request":"vault/vault","error":"failed to get certificate expiration: failed to parse certificate PEM","stacktrace":"github.com/go-logr/zapr.(*zapLogger).Error\n\t/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:217\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.1.10/pkg/internal/controller/controller.go:158\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:133\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:134\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/go/pkg/mod/k8s.io/apimachinery@v0.0.0-20181127025237-2b1284ed4c93/pkg/util/wait/wait.go:88"}
```

Has anyone found a way to disable the self signed cert?

**Comment by primeroz** [comment]
I Will try to get a full example with details tomorrow since now I am on mobile but the bottom line is you **do need** multiple listeners for this to work.

The one on port 8200 will use the secret created by the operator and it will be rotated by it. This is needed by all the other bits the operator does and changing it is more work than is worth it

The second listener, I run on port 8300, is set to use the secret created by cert manager and mounted in the vault pod by the operator using the *secret* field to the crd

Another field to the crd you will need is the *watchedsecrets* one which , using a label selector, need to match the secret created by cert manager. This way the operator will restart the vault pod when the secret is updated by cert manager

At this point have your external client reach the listener on port 8300 and you should be good. I use an ingress with SSL passthrough for that but a node port should work as well

**Comment by kunickiaj** [comment]
Yeah that's what I gathered. My fatal mistake was not using a separate port and letting the operator use its self signed one, which is a bit undesirable. A better long term fix would be to not hard code the filenames in the operator so that it could find the cert-manager managed ones as well. On the plus side, that means I can remove vault.vault and 127.0.0.1 from the SANs of the cert used for the exposed service. (Using GCP private Loadbalancer type service)

I'll move the cert-manager one over to another port for now, as you've suggested. I think I've got all the other bits working properly based on the https://github.com/banzaicloud/bank-vaults/blob/master/operator/deploy/cr-customports.yaml example.
